EduVPN-redesign/Controllers/SearchViewController.swift

```diff
@@ -7,21 +7,189 @@
+        // Walk the responder chain to find SearchViewController
+        var responder: NSResponder? = nextResponder
+        while responder != nil {
+            if let searchVC = responder as? SearchViewController {
+                searchVC.searchFieldGotFocus()
+                break
+            }
+            responder = responder?.nextResponder
+        }
+        return super.becomeFirstResponder()
+    }
+}
+
+// MARK: - Table view delegate and data source
+
+extension SearchViewController {
+    func numberOfRows() -> Int {
+        return viewModel?.numberOfRows() ?? 0
     }
+
+    func cellForRow(at index: Int, tableView: TableView) -> NSView? {
+        let row = viewModel.row(at: index)
+        if row.rowKind.isSectionHeader {
+            let cell = tableView.dequeue(SectionHeaderCell.self,
+                                         identifier: "SearchSectionHeaderCell",
+                                         indexPath: IndexPath(item: index, section: 0))
+            cell.configure(as: row.rowKind)
+            return cell
+        } else if row.rowKind == .noResultsKind {
+            let cell = tableView.dequeue(SearchNoResultsCell.self,
+                                         identifier: "SearchNoResultsCell",
+                                         indexPath: IndexPath(item: index, section: 0))
+            return cell
+        } else {
+            let cell = tableView.dequeue(RowCell.self,
+                                         identifier: "SearchRowCell",
+                                         indexPath: IndexPath(item: index, section: 0))
+            cell.configure(with: row)
+            return cell
+        }
+    }
+
+    func canSelectRow(at index: Int) -> Bool {
+        return viewModel.row(at: index).rowKind.isServerRow
+    }
+
+    func didSelectRow(at index: Int) {
+        let row = viewModel.row(at: index)
+        let serverAuthService = parameters.environment.serverAuthService
+        let navigationController = parameters.environment.navigationController
+        let delegate = self.delegate
+        if let baseURL = row.baseURL {
+            navigationController?.showAuthorizingMessage(cancelAuthorizationHandler: {
+                serverAuthService.cancelAuth()
+                navigationController?.hideAuthorizingMessage()
+            })
+            firstly {
+                serverAuthService.startAuth(baseURL: baseURL, from: self)
+            }.ensure {
+                makeApplicationComeToTheForeground()
+                navigationController?.hideAuthorizingMessage()
+            }.map { authState in
+                delegate?.searchViewControllerAddedServer(baseURL: baseURL, authState: authState)
+                navigationController?.popViewController(animated: true)
+            }.catch { error in
+                if !serverAuthService.isUserCancelledError(error) {
+                    navigationController?.showAlert(for: error)
+                }
+            }
+        }
```

Comment:
I am still not happy with this approach at all. Now we still are having logic with regards to how the flow through the app works in several places throughout various view controllers instead of having that in separate coordinators are per my suggested app architecture. This search view controller for example knows way too much about other parts of the app that it shouldn't.

A coordinator can manage multiple view controllers that are part of the same section of the app instead of having to split that over the view controllers. With your approach splitting up a part of the app into multiple screens becomes really difficult. Also, the view controller here is handling what gets shown and hidden and presented and dismissed making for the view controller to have to know for certain that it can do that safely.  

I can not in good conscience approve this pull request.

Reply:
> This search view controller for example knows way too much about other parts of the app that it shouldn't.

Yes, it does know some things about other parts of the app. Here, it assumes that (1) there exists a navigation controller, and (b) it has been pushed using that navigation controller.

In your scheme, this assumption would be done by the search coordinator, right? If yes, why is the search coordinator making the assmption better than the search vc making that assumption?